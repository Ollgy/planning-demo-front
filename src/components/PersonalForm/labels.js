const labels = {
  lastName: 'Фамилия',
  firstName: 'Имя',
  middleName: 'Отчество',
  position: 'Должность',
  birthDate: 'Дата рождения',
  phone: 'Телефон',
  email: 'Эл.почта',
  messengers: 'Мессенджеры'
};

const placeholders = {
  lastName: '. . .',
  firstName: '. . .',
  middleName: '. . .',
  position: '. . .',
  birthDate: 'дд.мм.гггг',
  phone: '. . .',
  email: '. . .',
  messengers: '. . .'
};

export { labels, placeholders };
