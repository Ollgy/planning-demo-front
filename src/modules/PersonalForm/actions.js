import { createAction } from 'redux-actions';
import api from '../../api';

export const setFields = createAction('SET_FIELDS_PERSONAL');
export const setViewMode = createAction('SET_VIEW_MODE_PERSONAL');
export const setDialog = createAction('SET_DIALOG_PERSONAL');
export const validateForm = createAction('VALIDATE_FORM_PERSONAL');
export const inputForm = createAction('INPUT_FORM_PERSONAL');
export const clearError = createAction('CLEAR_ERROR_PERSONAL');

export const getPositionsRequest = createAction('GET_POSITIONS_REQUEST');
export const getPositionsSuccess = createAction('GET_POSITIONS_SUCCESS');
export const getPositionsFailure = createAction('GET_POSITIONS_FAILURE');

export const getPositions = () => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(getPositionsRequest());

      api
        .getPositions()
        .then(positions => {
          dispatch(getPositionsSuccess(positions));
          resolve();
        })
        .catch(error => {
          dispatch(getPositionsFailure(error));
          reject();
        });
    });
};

export const saveNewUserRequest = createAction('SAVE_NEW_USER_REQUEST');
export const saveNewUserSuccess = createAction('SAVE_NEW_USER_SUCCESS');
export const saveNewUserFailure = createAction('SAVE_NEW_USER_FAILURE');

export const saveNewUser = userData => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(saveNewUserRequest());

      api
        .saveNewUser(userData)
        .then(user => {
          dispatch(saveNewUserSuccess(user));
          resolve();
        })
        .catch(error => {
          dispatch(saveNewUserFailure(error));
          reject();
        });
    });
};

export const updateUserRequest = createAction('UPDATE_USER_REQUEST');
export const updateUserSuccess = createAction('UPDATE_USER_SUCCESS');
export const updateUserFailure = createAction('UPDATE_USER_FAILURE');

export const updateUser = (userId, userData) => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(updateUserRequest());

      api
        .updateUser(userId, userData)
        .then(user => {
          dispatch(updateUserSuccess(user));
          resolve();
        })
        .catch(error => {
          dispatch(updateUserFailure(error));
          reject();
        });
    });
};

export const saveUserImageRequest = createAction('UPDATE_USER_REQUEST');
export const saveUserImageSuccess = createAction('UPDATE_USER_SUCCESS');
export const saveUserImageFailure = createAction('UPDATE_USER_FAILURE');

export const saveUserImage = (userId, imgData) => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(saveUserImageRequest());

      api
        .saveUserImage(userId, imgData)
        .then(user => {
          dispatch(saveUserImageSuccess(user));
          resolve();
        })
        .catch(error => {
          dispatch(saveUserImageFailure(error));
          reject();
        });
    });
};
