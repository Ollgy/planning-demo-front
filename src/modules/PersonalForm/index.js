export { default } from './reducerMain';
export * from './actions';
