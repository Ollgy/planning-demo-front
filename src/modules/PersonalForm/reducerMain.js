import { combineReducers } from 'redux';
import positions from './reducers/positions';
import userNew from './reducers/userNew';
import userUpdate from './reducers/userUpdate';
import userImage from './reducers/userImage';
import datalist from './reducers/datalist';
import view from './reducers/view';
import dialog from './reducers/dialog';

export default combineReducers({
  positions,
  datalist,
  view,
  dialog,
  userUpdate,
  userNew,
  userImage
});
