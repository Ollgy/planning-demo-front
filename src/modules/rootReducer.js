import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
