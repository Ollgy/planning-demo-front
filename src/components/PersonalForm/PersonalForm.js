import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import Button from '../Button';
import PersonalPhoto from '../PersonalPhoto';
import Input from '../Input';
import InputDataList from '../InputDataList';
import InputDate from '../InputDate';
import Textarea from '../Textarea';
import FormMsg from '../FormMsg';
import Preloader from '../Preloader';

import { labels, placeholders } from './labels';
import prepareText from './utils/prepareText';
import {
  getPositions,
  setFields,
  setViewMode,
  setDialog,
  inputForm,
  validateForm,
  clearError,
  saveUserImage,
  saveNewUser,
  updateUser
} from '../../modules/PersonalForm';

import { M, D, F, P } from '../../const';
import styles from '../Profile/Profile.module.css';
import formStyles from '../LoginForm/LoginForm.module.css';

const readOnly = ['firstName', 'lastName', 'middleName', 'email', 'position'];

class PersonalForm extends PureComponent {
  constructor(props) {
    super(props);
    this.activeInput = React.createRef();
  }

  componentWillUnmount() {
    this.props.setFields();
  }

  async componentDidMount() {
    const {
      user,
      startMode,
      getPositions,
      setFields,
      setViewMode
    } = this.props;

    setViewMode(startMode);

    await getPositions();

    if (startMode !== M.create && user) {
      setFields({ ...user, position: user.position.name });
    } else {
      setFields();
    }

    if (this.activeInput.current) {
      this.activeInput.current.focus();
    }
  }

  onInputChange = e => {
    const { inputForm, clearError } = this.props;
    const { value, name } = e.target;

    inputForm({ input: name, value });
    clearError({ input: name });
  };

  onInputSelect = (positions, e) => {
    const { setFields, clearError } = this.props;
    const { value, name } = e.target;

    if (positions.includes(value)) {
      setFields({ [name]: value });
      clearError({ input: name });
    }
  };

  onUpdateMode = e => {
    e.preventDefault();

    this.props.setViewMode(M.update);

    if (this.activeInput.current) {
      this.activeInput.current.focus();
    }
  };

  onPhotoUpload = e => {
    const { setFields } = this.props;
    const input = e.target;
    const reader = new FileReader();

    reader.onload = e => setFields({ image: e.target.result });

    if (input.files[0]) {
      reader.readAsDataURL(input.files[0]);
    }
  };

  onSetDate = async (strDate, field) => {
    const { setFields, setDialog } = this.props;

    setFields({ [field]: strDate });
    setDialog({ name: D.calendar, status: false });
  };

  onSubmit = async e => {
    e.preventDefault();

    const form = e.target;
    const mode = form.name;
    const imgUpload = form.querySelector('#file').files[0];
    const {
      fields,
      positions,
      user = {},
      validateForm,
      setViewMode,
      saveNewUser,
      saveUserImage,
      updateUser
    } = this.props;

    await validateForm({ fields, positions });

    const { validation } = this.props;

    const updateFields = Object.keys(fields).filter(field =>
      field === F.position
        ? fields[field] !== (user[field] ? user[field].name : null)
        : fields[field] !== user[field] && field !== F.image
    );

    const pos = positions.find(pos => pos.name === fields.position);

    if (validation.status && mode === M.create) {
      await saveNewUser({
        ...fields,
        position: { id: pos.id, name: pos.name, permission: pos.permission },
        image: ''
      });
    } else if (validation.status && mode === M.update && updateFields.length) {
      const updateData = updateFields.reduce(
        (obj, field) => ({ ...obj, [field]: fields[field] }),
        {}
      );

      if (updateData.position) {
        updateData.position = {
          id: pos.id,
          name: pos.name,
          permission: pos.permission
        };
      }

      const { userNew } = this.props;

      await updateUser(user.id ? user.id : userNew.id, updateData);
    }

    const { userNew } = this.props;

    if (imgUpload) {
      await saveUserImage(userNew ? userNew.id : user.id, new FormData(form));
    }

    validation.status ? setViewMode(M.read) : this.activeInput.current.focus();
  };

  render() {
    const {
      labels,
      permission,
      fields,
      validation,
      positions,
      mode,
      dialog,
      isLoading,
      setDialog
    } = this.props;
    const { msg, error, errorInputs } = validation;

    return !isLoading ? (
      <form
        action="/"
        encType="multipart/form-data"
        className={styles.formData}
        onSubmit={this.onSubmit}
        name={mode}
      >
        <div className={styles.infoWrap}>
          <div className={styles.infoBlock}>
            {Object.keys(fields)
              .filter(field => field !== F.image)
              .map((field, i) => (
                <div className={formStyles.formItem} key={field}>
                  {mode === M.read ? (
                    <span
                      className={styles.formValue}
                      dangerouslySetInnerHTML={{
                        __html: fields[field]
                          ? prepareText(field, fields[field])
                          : ''
                      }}
                    />
                  ) : field === F.birthDate ? (
                    <InputDate
                      type="text"
                      name={field}
                      ref={errorInputs[0] === field ? this.activeInput : null}
                      value={fields[field]}
                      readOnly={true}
                      placeholder={placeholders[field]}
                      errorClass={errorInputs.includes(field) ? error : null}
                      onFocus={setDialog.bind(null, {
                        name: D.calendar,
                        status: true
                      })}
                      isOpen={dialog.calendar}
                      activities={{
                        openDialog: setDialog.bind(null, {
                          name: D.calendar,
                          status: true
                        }),
                        closeDialog: setDialog.bind(null, {
                          name: D.calendar,
                          status: false
                        }),
                        setDate: this.onSetDate
                      }}
                    />
                  ) : field === F.position ? (
                    <InputDataList
                      name={field}
                      ref={errorInputs[0] === field ? this.activeInput : null}
                      value={fields[field]}
                      placeholder={placeholders[field]}
                      readOnly={
                        mode !== M.create &&
                        permission !== P.manager &&
                        readOnly.includes(field)
                      }
                      errorClass={errorInputs.includes(field) ? error : null}
                      list={positions}
                      onChange={this.onInputChange}
                    />
                  ) : field === F.messengers ? (
                    <Textarea
                      ref={errorInputs.length ? null : this.activeInput}
                      name={field}
                      value={fields[field]}
                      placeholder={placeholders[field]}
                      onChange={this.onInputChange}
                      onFocus={setDialog.bind(null, {
                        name: D.calendar,
                        status: false
                      })}
                      errorClass={errorInputs.includes(field) ? error : null}
                    />
                  ) : (
                    <Input
                      type="text"
                      ref={
                        errorInputs.length
                          ? field === errorInputs[0]
                            ? this.activeInput
                            : null
                          : null
                      }
                      name={field}
                      value={fields[field]}
                      readOnly={
                        mode !== M.create &&
                        permission !== P.manager &&
                        readOnly.includes(field)
                      }
                      placeholder={placeholders[field]}
                      onChange={this.onInputChange}
                      onFocus={setDialog.bind(null, {
                        name: D.calendar,
                        status: false
                      })}
                      errorClass={errorInputs.includes(field) ? error : null}
                    />
                  )}
                  <label className={formStyles.formLabel} htmlFor={field}>
                    <span>{labels[field]}</span>
                  </label>
                </div>
              ))}
          </div>
          <div className={styles.msgBlock}>
            <FormMsg children={msg} />
          </div>
          <div className={styles.buttonBlock}>
            {mode === M.read ? (
              <Button
                type="button"
                name="update"
                onClick={this.onUpdateMode}
                children="Редактировать"
              />
            ) : mode === M.create ? (
              <Button
                type="submit"
                name="create"
                children="Добавить пользователя"
              />
            ) : (
              <Button type="submit" name="save" children="Сохранить" />
            )}
          </div>
        </div>
        <div className={styles.photoBlock}>
          <PersonalPhoto
            image={fields.image}
            mode={mode}
            onPhotoUpload={this.onPhotoUpload}
          />
        </div>
      </form>
    ) : (
      <Preloader />
    );
  }
}

const withData = labels => Component => {
  const wrap = props => <Component {...props} labels={labels} />;

  wrap.displayName = `WithData(${getDisplayName(Component)})`;

  return wrap;
};

const getDisplayName = WrappedComponent => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
};

const mapStateToProps = state => ({
  positions: state.personalForm.positions.items,
  fields: state.personalForm.datalist.fields,
  validation: state.personalForm.datalist.validation,
  mode: state.personalForm.view.mode,
  dialog: state.personalForm.dialog.data,
  isLoading:
    state.personalForm.positions.isLoading ||
    state.personalForm.userNew.isLoading ||
    state.personalForm.userUpdate.isLoading ||
    state.personalForm.userImage.isLoading,
  userNew: state.personalForm.userNew.user
});

const mapDispatchToProps = dispatch => ({
  getPositions: () => dispatch(getPositions()),
  saveNewUser: arg => dispatch(saveNewUser(arg)),
  updateUser: (arg1, arg2) => dispatch(updateUser(arg1, arg2)),
  saveUserImage: (arg1, arg2) => dispatch(saveUserImage(arg1, arg2)),
  setFields: arg => dispatch(setFields(arg)),
  setViewMode: arg => dispatch(setViewMode(arg)),
  setDialog: arg => dispatch(setDialog(arg)),
  inputForm: arg => dispatch(inputForm(arg)),
  validateForm: arg => dispatch(validateForm(arg)),
  clearError: arg => dispatch(clearError(arg))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withData(labels)(PersonalForm));
