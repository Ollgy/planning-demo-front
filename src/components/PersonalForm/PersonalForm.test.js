import React from 'react';
import PersonalForm from './PersonalForm';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<PersonalForm/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<PersonalForm store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<PersonalForm store={store} />);
  const inner = wrapper.children().first();

  it('contains fields in props', () => {
    expect(inner.props().fields).toBeTruthy();
  });

  it('contains positions in props', () => {
    expect(inner.props().positions).toBeTruthy();
  });
});
