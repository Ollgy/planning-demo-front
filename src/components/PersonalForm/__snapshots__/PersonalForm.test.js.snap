// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PersonalForm/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <WithData(PersonalForm)
    clearError={[Function]}
    dialog={
      Object {
        "calendar": false,
      }
    }
    fields={
      Object {
        "birthDate": "",
        "email": "",
        "firstName": "",
        "image": "",
        "lastName": "",
        "messengers": "",
        "middleName": "",
        "phone": "",
        "position": "",
      }
    }
    getPositions={[Function]}
    inputForm={[Function]}
    isLoading={false}
    mode="read"
    positions={Array []}
    saveNewUser={[Function]}
    saveUserImage={[Function]}
    setDialog={[Function]}
    setFields={[Function]}
    setViewMode={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    updateUser={[Function]}
    userNew={null}
    validateForm={[Function]}
    validation={
      Object {
        "error": "",
        "errorInputs": Array [],
        "msg": "",
        "status": false,
      }
    }
  />
</ContextProvider>
`;
